main.nf

```diff
@@ -303,26 +326,50 @@ if (error) exit 1, error
 
 if (warning) log.warn warning
 
-// Determine merge groups
+// Determine merge groups for mapping
+// ######################
+if (merge_mapping && metadata) {
+
+  metadata_to_merge_mapping.filter {
+    it.values()[0] in inputs.collect{it['prefix']}
+  }.map {
+    factors = it.subMap(merge_mapping).values()
+    id = '' in factors ? 'NA_' + it.values()[0] : factors.join('_')
+    [id, it.values()[0]]
+  }.groupTuple().into {
+    groups_to_merge_mapping
+    groups_to_log_mapping
+  }
+
+  groups_to_log_mapping = groups_to_log_mapping.toList().get().sort { a, b -> a[0] <=> b[0] }
+
+  s = groups_to_log_mapping.size() > 1 ? 's' : ''
+  info = "The following merge group$s will be created for mapping:\n  "
+  info += groups_to_log_mapping.collect{ it[0] + ': ' + it[1].join('  ') }.join('\n  ')
+
+  log.info info
+}
+
+// Determine merge groups for assembly
```

Comment:
Code duplication here. It's okay as a first draft. We can probably refactor this with Groovy's collect method, or eventually use Nextflow DSL2 to define a function.